### PR TITLE
refactor(update): share host update-check state machine

### LINKED ISSUE #9515: Consolidate duplicated update-checker state machines across CLI and desktop
## Context

Follow-up from #9506 (refactor: consolidate duplicated logic and native-method swaps).

The PR consolidated 8 duplicated patterns across the codebase, but one finding was explicitly deferred: `packages/cli/src/runtime/updateChecker.ts` and `packages/desktop/src/main/desktopUpdateChecker.ts` both independently implement the same state machine twice:

1. A daily-throttled "check remote version, notify, stamp `lastCheckedAt`" loop
2. A "fetch a version field over HTTP with a timeout, swallow failures" helper

## Details

Extract the shared state-machine core into a single location (likely `src/utils/` or a new shared module), parameterized by:
- Global-state key names (CLI and desktop use different persistence keys)
- Notify semantics (desktop shows a system notification vs. CLI prints to stderr)

Keep host-specific wiring (IPC/channels, notification surfaces, polling strategies) at the call sites so the shared core stays platform-agnostic.

## References

- PR: #9506

### PR DESCRIPTION
**Pull request by LionSR**
## Summary

- Centralize daily throttling, semantic-version comparison, notification ordering, and successful-check stamping in one host-neutral update-check module.
- Share timeout-bounded, failure-tolerant JSON version fetching between the CLI registry and desktop release sources.
- Retain host-specific behavior: CLI installation prompts and best-effort stamp writes, desktop one-notification-per-version state and in-flight window ownership.

## Issue acceptance gates

- #9515: one shared state-machine core accepts host-specific state keys and notification callbacks.
- #9515: one shared HTTP helper owns timeout, response, payload, and failure handling for string version fields.
- #9515: host-specific IPC, notification, interactive update, and concurrency behavior remains at each host boundary.
- Empty desktop release tags remain failed fetches and therefore do not suppress retries.
- Empty npm registry version fields are now likewise treated as failed refreshes; they are invalid semantic versions and no longer stamp the daily throttle.

## Single source of truth

`src/utils/system/updateCheck.ts` now owns the daily update-check lifecycle and resilient JSON-field fetch. `semverUpdateCheck.ts` remains the owner of semantic-version comparison, cadence, and the common opt-out environment variable.

## Duplication and abstraction budget

The CLI and desktop previously encoded the same throttle, fetch, comparison, notify, and stamp ordering independently. The new module has two production consumers and owns the cross-host invariant. It is not a pass-through: it hides retry eligibility, notification suppression, persistence ordering, and stamp-failure policy. Host adapters retain only genuine host differences.

## Net elements (R6)

- Files: 5 changed; 1 production file added; no files deleted.
- Exported symbols: 3 shared exports added (`UpdateCheckFetchResult`, `runDailyUpdateCheck`, `fetchJsonStringField`); the existing `CliUpdateFetchResult` export remains as a type alias.
- Types/functions: net +2 interfaces and +2 functions; no classes or enums added.
- LoC: +340/-130, net +210 overall. Production code is +188/-128, net +60; direct and regression tests account for the remaining net +150.
- Justification: the small production increase replaces two independent state machines and two independent HTTP failure handlers with one two-consumer owner. The test increase directly covers the shared policy branches and the empty-tag regression identified in review.

## Consumer counts (R8)

- `runDailyUpdateCheck`: 2 production consumers (CLI and desktop).
- `fetchJsonStringField`: 2 production consumers (CLI and desktop).
- `UpdateCheckFetchResult`: 1 production type consumer through the public CLI result alias.
- No emit path, subscriber key, or existing export was deleted or rerouted.

## Host impact

Extension: None.

Desktop: Uses the shared lifecycle and JSON fetch while retaining packaging checks, environment opt-out, in-flight coalescing, latest-window notification ownership, and per-release suppression.

CLI: Uses the shared lifecycle and JSON fetch while retaining TTY/source-install gates, npm/Homebrew source selection, interactive installation, and best-effort throttle persistence.

## Scheduled refactoring

None. The duplicate implementations are removed in this PR; no dual-system transition remains.

## Validation

- `npm run format`
- `npm run compile:fast`
- `npm run lint`
- `npm run typecheck`
- `npm test` (806 files passed, 1 skipped; 8176 tests passed, 4 skipped)
- Latest focused review-fix suite: 20 tests passed.
- `npm run typecheck:test-kernel`

Closes #9515

---

> [!NOTE]
> **Medium Risk**
> Touches startup update-check paths for CLI and desktop; behavior is intended to be equivalent but ordering and empty-tag handling are subtle enough to warrant careful review.
> 
> **Overview**
> Introduces **`src/utils/system/updateCheck.ts`** as the shared owner of daily-throttled update checks and timeout-bounded JSON version fetches, replacing parallel logic in the CLI and desktop checkers.
> 
> **`runDailyUpdateCheck`** centralizes the 24h throttle, semver comparison, notify-before-stamp ordering, optional once-per-release notification (`lastNotifiedVersionKey`), and the `refreshed` gate so stale sources (e.g. failed Homebrew tap refresh) do not suppress retries. **`fetchJsonStringField`** unifies npm registry and GitHub release fetches and now treats **empty string fields as no version** (e.g. empty `tag_name`).
> 
> CLI **`checkCliUpdateAvailable`** delegates to the shared helper with **`stampFailure: 'ignore'`** so read-only global state cannot undo an accepted update. Desktop **`checkForDesktopUpdate`** delegates while keeping packaging/env gates and in-flight coalescing at the host boundary.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 44b97feaedb4458edcd023052a11b75d1c7f6402. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/utils/system/updateCheck.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  state: StateStore;
+  lastCheckedAtKey: string;
+  /** Persisted version used by hosts that notify only once per release. */
+  lastNotifiedVersionKey?: string;
+  fetchLatest: () => Promise<UpdateCheckFetchResult>;
+  notify: (latest: string) => PromiseLike<void> | void;
+  now?: () => number;
+  /** Whether failure to persist the throttle stamp rejects the check. */
+  stampFailure?: 'throw' | 'ignore';
+}
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/utils/system/updateCheck.ts]: Removed both unused option-type exports in 8d5947f98a; they are now private implementation details of the shared module.

File: packages/desktop/src/main/desktopUpdateChecker.ts [hunk 2]
```diff
@@ -43,23 +41,17 @@ async function fetchLatestDesktopRelease(options?: {
...
+    url: RELEASES_API_URL,
+    field: 'tag_name',
+    timeoutMs: options?.timeoutMs ?? FETCH_TIMEOUT_MS,
+    headers: {
+      accept: 'application/vnd.github+json',
+      'user-agent': GITHUB_USER_AGENT,
+    },
+    fetchImpl: options?.fetchImpl,
+  });
+  return tag === undefined ? undefined : { version: tag.replace(/^v/, '') };
```
[the remark this answers is not in the document]
Comment (LionSR) [reply packages/desktop/src/main/desktopUpdateChecker.ts]: Restored the non-empty tag boundary in 8d5947f98a and added a desktop regression test proving an empty tag leaves the daily stamp unset.

**Comment by LionSR** [comment]
@codex review

File: src/utils/system/updateCheck.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+/** Result of consulting an update source. */
+export interface UpdateCheckFetchResult {
+  /** Published version, when the source supplied one. */
+  version: string | undefined;
+  /** Whether the result came from a successful live refresh. */
+  refreshed: boolean;
+}
+
+export interface DailyUpdateCheckOptions {
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/utils/system/updateCheck.ts]: The option interfaces are private as of 8d5947f98a. This duplicate earlier suggestion is now addressed.

File: src/utils/system/updateCheck.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      signal: AbortSignal.timeout(timeoutMs),
+      headers,
+    });
+    if (!response.ok) return undefined;
+
+    const body: unknown = await response.json();
+    if (typeof body !== 'object' || body === null) return undefined;
+    const value = (body as Record<string, unknown>)[field];
+    return typeof value === 'string' ? value : undefined;
+  } catch {
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/utils/system/updateCheck.ts:117]: Added the requested inline explanation in 5bc7950e7b, naming timeout, network, and invalid-JSON failures.

**Comment by LionSR** [comment]
@codex review

**Comment by LionSR** [comment]
@codex review

**Comment by LionSR** [comment]
Addressed the final TeXRA review in 44b97feaed: the shared JSON-field helper now rejects empty strings directly, and the best-effort stamp branch now catches both synchronous exceptions and rejected writes. Direct regression tests cover both cases; focused tests, formatting, lint, and test-kernel type checking pass.

**Comment by LionSR** [comment]
Final TeXRA review observations addressed: the PR body now explicitly records the intentional CLI edge-case change for empty registry versions. No mock cleanup change is needed: each `spyOn` targets a fresh `FakeStateStore` instance local to one test, so it cannot affect another test or suite.
